to-json-schema.js

```diff
@@ -136,21 +136,39 @@ function validateAvroSchema(avroDefinition) {
   avsc.Type.forSchema(avroDefinition);
 }
 
-async function convertAvroToJsonSchema(avroDefinition, isTopLevel) {
+/**
+ * Cache the passed value under the given key. If the key is undefined the value will not be cached. This function
+ * uses mutation of the passed cache object rather than a copy on write cache strategy.
+ *
+ * @param cache Map<String, JsonSchema> the cache to store the JsonSchema
+ * @param key String | Undefined - the fully qualified name of an avro record
+ * @param value JsonSchema - The json schema from the avro record
+ */
+function cacheAvroRecordDef(cache, key, value) {
+  if (key) {
+    cache[key] = value;
+  }
+}
+
+async function convertAvroToJsonSchema(avroDefinition, isTopLevel, recordCache = {}) {
   const jsonSchema = {};
   const isUnion = Array.isArray(avroDefinition);
 
-  validateAvroSchema(avroDefinition);
```

Comment:
This was moved outside the recursion because in the case of union Avro schemas the entire schema is valid but the sub pieces are not when the `n + 1` records contain a type defined in `n`